[Experimental] Prevent string-based sanitization on non-string field types in additional checkout fields

opr opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

#### Main changes

- Sanitize address fields based on their schema type.
- Add a new class `SanitizationUtils` which contains a `wp_kses_array` function. After all sanitization is complete in the `AbstractAddressSchema` and `CheckoutSchema`, do a final pass over the array recursively running `wp_kses` on any string key.

#### Required changes because of the above

- Update the `BillingAddressSchema` and `AbstractAddressSchema` to run _validation_ before sanitization. This is required because of the main change.
  - In `BillingAddressSchema`, the address was sanitized in the `validate_callback` function. This meant that things were sanitized _before_ validation had finished. The issue was that this could potentially result in casting invalid values (e.g. arrays) to strings. The `BillingAddressSchema` does not need to sanitize fields before validating them because it only validates the email, which had no special sanitization in place anyway.
  - In `AbstractAddressSchema` we still need to sanitize before validation because we want the keys to be normalized, e.g. state to upper case. however we need to prevent values being cast incorrectly. To achieve this, each key is validated against its schema definition first, then sanitized after they all pass, then validated.

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

```php
add_action(
	'woocommerce_blocks_loaded',
	function() {

		woocommerce_blocks_register_checkout_field(
			array(
				'id'       => 'plugin-namespace/email-opt-in',
				'label'    => 'Sign up to our mailing list?',
				'location' => 'contact',
				'type'     => 'checkbox',
			),
		);
    }
);
```

1. Add a checkout field to your store using the snippet above
2. Add an item to your cart, then go to the Checkout block.
4. Check and uncheck the box
5. Place the order succesfully.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->
Skipping changelog entry as this is behind a feature flag.

</details>

opr committed:
Skip wp_kses unless field is of type string

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

senadir commented on plugins/woocommerce/src/StoreApi/Schemas/V1/CheckoutSchema.php:
```diff
@@ -339,7 +339,12 @@ function( $carry, $key ) use ( $fields, $properties ) {
-				$carry[ $key ]  = wp_kses( $rest_sanitized, [] );
+				$carry[ $key ]  = $rest_sanitized;
+
+				// Specific sanitization for string types, skipping other types from being coerced to strings.
+				if ( 'string' === $field_schema['type'] ) {
+					$carry[ $key ] = wp_kses( $rest_sanitized, [] );
+				}
```
Should this be done in the address as well?

opr replied:
Added in c283b95df2ed0038bd4e76cfddbb42f02e46a4b0

opr committed:
wp_kses on address fields of type string

mikejolley commented on plugins/woocommerce/src/StoreApi/Schemas/V1/AbstractAddressSchema.php:
```diff
@@ -130,6 +131,10 @@ function( $carry, $key ) use ( $address, $validation_util ) {
 						$carry[ $key ] = $address[ $key ];
+						// Specific sanitization for string types, skipping other types from being coerced to strings.
+						if ( 'string' === $field_schema[ $key ]['type'] ) {
+							$carry[ $key ] = wp_kses( $address[ $key ], [] );
```
If this is not a string, is the value still sanitised somehow? I see the other method uses `rest_sanitize_value_from_schema`.

opr replied:
True, not sure why they're different. I added `rest_sanitize_value_from_schema` to this for values we don't already handle (`country`, `state`, and `postcode`)

opr committed:
Sanitize fields from the schema

wavvves commented:
Hi @opr I see by default all were strings, do you think there might be other schemas that were also all strings that might need something similar?

There are some[ tests failing](https://github.com/woocommerce/woocommerce/actions/runs/7642165597/job/20821205716?pr=43999) that seem related 😖

opr committed:
Get correct field from schema

opr commented:
Hi @wavvves thanks for the review 

> do you think there might be other schemas that were also all strings that might need something similar?

You mean other schemas that cast field values to string? Possibly, yes, but I didn't look into that because it's out of scope for this PR and currently isn't presenting any issues (no bug reports etc.)

If you mean something else then please could you clarify? Thanks!

opr committed:
Use sanitized string in wp_kses

wavvves commented:
> Hi @wavvves thanks for the review
> 
> > do you think there might be other schemas that were also all strings that might need something similar?
> 
> You mean other schemas that cast field values to string? Possibly, yes, but I didn't look into that because it's out of scope for this PR and currently isn't presenting any issues (no bug reports etc.)
> 
> If you mean something else then please could you clarify? Thanks!

My bad I was looking at the ImageAttachmentSchema that is also part of the CheckoutSchema, in which one of the properties is an integer, but yeah, it doesn't apply here. 
[Unit tests ](https://github.com/woocommerce/woocommerce/actions/runs/7667734357/job/20898225356?pr=43999) are still failing, but I'm doing the testing steps ahead 👍🏼

opr committed:
Don't sanitize billing address before validating email

This will prevent sanitization errors being encountered if there are any validation errors, validation should run first and catch these before they make it to sanitize

opr committed:
Validate each address field before sanitzation

opr committed:
Add SanitizationUtils class to handle kses on arrays

opr committed:
Handle passing non-array types

opr committed:
Add unit tests for SanitizationUtils

opr committed:
Run wp_kses on addresses/checkout payloads

opr committed:
Remove check and use type in function definition to ensure array

opr committed:
Remove incorrect type test

opr committed:
Add nested bad string to test

senadir commented on plugins/woocommerce/src/StoreApi/Schemas/V1/AbstractAddressSchema.php:
```diff
@@ -152,9 +156,32 @@ function( $carry, $key ) use ( $address, $validation_util ) {
+		$address         = (array) $address;
 		$validation_util = new ValidationUtils();
 
+		// Validate each key manually before sanitizing. The reason we need to sanitize "before" validation is to ensure
```
This comment is confusing.

You're validating before sanitizing.

opr replied:
Ah yeah OK i can improve this.

Basically the flow is Validate -> Sanitize -> Validate

1st validate is to ensure fields match their schema, sanitize to put them in the correct format, then 2nd validate to ensure the correctly-formatted values match what we expect (postcode etc.)

senadir reviewed (approved):
Testing fine for me.

wavvves commented on plugins/woocommerce/src/StoreApi/Utilities/SanitizationUtils.php:
```diff
@@ -0,0 +1,27 @@
+	 * @return mixed       The array, all string keys will have been run through wp_kses.
+	 */
+	public function wp_kses_array( array $array ) {
+		foreach ( $array as $key => $value ) {
```
We might wanna skip empty values with this, or adding to the nested `ifs` to skip recursiveness/calling unnecessary `wp_kses()`

```suggestion
		foreach ( $array as $key => $value ) {
		   if ( empty( $value ) {
		       continue;
		   }
```

wavvves commented on plugins/woocommerce/src/StoreApi/Utilities/SanitizationUtils.php:
```diff
@@ -0,0 +1,27 @@
+				$array[ $key ] = $this->wp_kses_array( $value );
+			}
+			if ( is_string( $value ) ) {
+				$array[ $key ] = wp_kses( $value, [] );
```
Going through the rabbit hole of `wp_kses()` (heavily recommended), at some point it resorts to [`preg_replace_callback()`](https://www.php.net/manual/en/function.preg-replace-callback.php) which might return `null` on errors. By adding the suggested code above, we can check for `null` here (original `null` values would have been caught and skipped already) and handle this.
Not saying we need it, but would it be ok stripping stuff out and sending `null` in it's place? 🤔

opr replied:
Check for null after the wp_kses has run? I think if the string is cleaned and something causes it to be null then it's fine to return null, something else went even more wrong.

wavvves replied:
Fair enough, resolving and approving this 👍

wavvves commented on plugins/woocommerce/tests/php/src/Blocks/StoreApi/Utilities/SanitizationUtilTest.php:
```diff
@@ -0,0 +1,48 @@
+	/**
+	 * @testdox `kses_array` should return an array of the same shape, but strings should be cleaned with `wp_kses`.
+	 */
+	public function test_kses_array() {
```
Once again, looking at the [`wp_kses()`](https://developer.wordpress.org/reference/functions/wp_kses/) source we might want to test sending heavily, beyond any salvation, malformed HTML 👿

opr replied:
Feels like that test is more focused on expecting `wp_kses` to behave a certain way no? I've added a test with a more mangled HTML string but I don't know if we should go further than that.

wavvves replied:
Maybe you are right, and the scope of testing falls a bit outside what the PR introduces. Nevertheless thanks for adding the edge case 💪

opr committed:
Improve comment for clarity

opr committed:
Add further tests

opr committed:
Skip/preserve empty values

opr committed:
Additional edge-case tests

mikejolley commented on plugins/woocommerce/src/StoreApi/Schemas/V1/AbstractAddressSchema.php:
```diff
@@ -152,9 +156,34 @@ function( $carry, $key ) use ( $address, $validation_util ) {
+		// correct format, and finally the second validation step is to ensure the correctly-formatted values
+		// match what we expect (postcode etc.).
+		foreach ( $address as $key => $value ) {
+			if ( is_wp_error( rest_validate_value_from_schema( $value, $this->get_properties()[ $key ], $key ) ) ) {
```
Would it be more efficient to put `$this->get_properties()` in a variable rather than keep calling the method?

opr committed:
Save schema to variable instead of getting every loop

wavvves reviewed (approved):
Looks good Thomas!

Please check [this minor lint issue](https://github.com/woocommerce/woocommerce/actions/runs/7741192312/job/21107832111?pr=43999) before merging 👍🏼

opr committed:
Add spaces around array key
